pkg/nsx/services/vpc/vpc.go

```diff
@@ -471,3 +523,236 @@ func (s *VPCService) Cleanup() error {
+		log.Error(err, "create avi rule", "rule", newrule)
+		return err
+	}
+	nsxrule, err := service.NSXClient.VPCRuleClient.Get(orgId, projectId, *vpc.Id, spId, *newrule.Id)
+	if err != nil {
+		log.Error(err, "get avi rule", "rule", nsxrule)
+		return err
+	}
+	service.RuleStore.Add(nsxrule)
+	log.Info("created avi rule")
+	return nil
+}
+
+func (service *VPCService) getorCreateAVIGroup(orgId string, projectId string, vpcId string, groupId string) (*model.Group, error) {
+	group, err := service.getAVIGroup(orgId, projectId, vpcId, groupId)
+	if err != nil {
+		log.Info("create avi group", "group", group)
+		group, err = service.createAVIGroup(orgId, projectId, vpcId, groupId)
+		if err != nil {
+			log.Error(err, "failed to create avi group", "group", groupId)
+		}
+	}
+	return group, err
+}
+
+func (service *VPCService) createAVIGroup(orgId string, projectId string, vpcId string, groupId string) (*model.Group, error) {
+	var group model.Group
+	tags := []model.Tag{
+		{
+			Scope: pointy.String(common.TagScopeCluster),
+			Tag:   pointy.String(service.NSXConfig.Cluster),
+		},
+		{
+			Scope: pointy.String(common.TagScopeVersion),
+			Tag:   pointy.String(strings.Join(common.TagValueVersion, ".")),
+		},
+		{
+			Scope: pointy.String(common.TagScopeVPCCRUID),
+			Tag:   pointy.String(vpcId),
+		},
+		{
+			Scope: pointy.String(common.TagScopeGroupType),
+			Tag:   pointy.String("avi"),
+		},
+	}
+	group.Tags = tags
+	expression := service.buildExpression("Condition", "VpcSubnet", "AVI_SUBNET_LB|", "Tag", "EQUALS", "EQUALS")
+	group.Expression = []*data.StructValue{expression}
+	group.DisplayName = pointy.String(groupId)
+
+	err := service.NSXClient.VpcGroupClient.Patch(orgId, projectId, vpcId, groupId, group)
+	if err != nil {
+		return &group, err
+	}
+	nsxgroup, err := service.NSXClient.VpcGroupClient.Get(orgId, projectId, vpcId, groupId)
+	return &nsxgroup, err
+}
+
+func (service *VPCService) buildExpression(resource_type, member_type, value, key, operator, scope_op string) *data.StructValue {
+	return data.NewStructValue(
+		"",
+		map[string]data.DataValue{
+			"resource_type":  data.NewStringValue(resource_type),
+			"member_type":    data.NewStringValue(member_type),
+			"value":          data.NewStringValue(value),
+			"key":            data.NewStringValue(key),
+			"operator":       data.NewStringValue(operator),
+			"scope_operator": data.NewStringValue(scope_op),
+		},
+	)
+}
+
+func (service *VPCService) buildAVIAllowRule(obj *model.Vpc, externalCIDR []string, groupId, ruleId, projectId string) (*model.Rule, error) {
+	rule := &model.Rule{}
+	rule.Action = pointy.String(model.Rule_ACTION_ALLOW)
+	rule.Direction = pointy.String(model.Rule_DIRECTION_IN_OUT)
+	rule.Scope = append(rule.Scope, groupId)
+	rule.SequenceNumber = pointy.Int64(math.MaxInt32 - 1)
+	rule.DestinationGroups = externalCIDR
+	rule.SourceGroups = append(rule.SourceGroups, "Any")
+	name := fmt.Sprintf("PROJECT-%s-VPC-%s-%s", projectId, *obj.Id, ruleId)
```

Comment:
need to have a check if it's meeting our naming convention for resource.

Reply:
no. It uses the nsxt naming style as we discussed before